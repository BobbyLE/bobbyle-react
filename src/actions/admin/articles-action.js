import database from '../../firebase/firebase';


//ADD_ARTICLE
export const addArticle = (article) => ({
  type: 'ADD_ARTICLE',
  article
});

export const startAddArticle = (articleData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      categoryId = '',
      body = '',
      published = false
    } = articleData;
    const article = { 
      title, 
      categories : {
        [categoryId]: true
      },
      body,
      published
    };

    // Generate new push ID for the new article
    const newArticleKey = database.ref(`users/${uid}`).child('articles').push().key;

    // Write the new article's data simultaneously in the articles list and the category's article list
    let updates = {};
    updates[`users/${uid}/categories/${categoryId}/articles/${newArticleKey}`] = true;
    updates[`users/${uid}/articles/${newArticleKey}`] = article;

    return database.ref().update(updates).then(() => {
      dispatch(addArticle({
        id: newArticleKey,
        ...article
      }));
    });
  };
};

// EDIT_ARTICLE
export const editArticle =  (id, updates) => ({
  type: 'EDIT_ARTICLE',
  id,
  updates
})

export const startEditArticle = (id, article) => {
  const updatedArticle = {
    title: article.title,
    categories: {
      [article.categoryId]: true
    },
    body: article.body,
    published: article.published
  }
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    
    //Edit article and simultaneously remove article from the old categories and add to new categories
    let updates = {}
    //Update article
    updates[`users/${uid}/articles/${id}`] = updatedArticle;
    //Remove article ID in each categories
    database.ref(`users/${uid}/categories`).orderByChild('articles')
      .once('value')
      .then((snapshot) => {
        snapshot.forEach( (childSnapshot) => {
          const key = childSnapshot.key;
          const val = childSnapshot.val();
          if(val.articles) {
            if(id in val.articles) {
              updates[`users/${uid}/categories/${key}/articles/${id}`] = null;  
            }
          }
        });
      }).then( () => {
        // Add article id to category articles list
        updates[`users/${uid}/categories/${article.categoryId}/articles/${id}`]  = true
        //Finally Query update
        return database.ref().update(updates).then(() => {
          return dispatch(editArticle(id, article));
        });
      })
  };
};


// REMOVE_ARTICLE
export const removeArticle = ({id} = {}) => ({
  type: 'REMOVE_ARTICLE',
  id
});
export const startRemoveArticle = ({id} = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    // Remove both articles and category articles
    let updates = {}
    // Remove the article item
    updates[`users/${uid}/articles/${id}`] = null;
    database.ref(`users/${uid}/categories`).orderByChild('articles')
      .once('value')
      .then((snapshot) => {
        snapshot.forEach( (childSnapshot) => {
          const key = childSnapshot.key;
          const val = childSnapshot.val();
          if(val.articles) {
            if(id in val.articles) {
              // Add updates to remove articles in each categories
              updates[`users/${uid}/categories/${key}/articles/${id}`] = null;  
            }
          }
        });
      }).then( () => {
        //Finally Query update delete
        return database.ref().update(updates).then(() => {
          return dispatch(removeArticle({id}));
        });
      })
  } 
};

// SET_ARTICLES
export const setArticles = (articles) => ({
  type: 'SET_ARTICLES',
  articles
});

export const startSetArticles = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid ? getState().auth.uid : process.env.USER_ID;
    return database.ref(`users/${uid}/articles`).once('value').then( (snapshot) => {
      const articles = [];
      snapshot.forEach( (childSnapshot) => {
        articles.push({
          id: childSnapshot.key,
          ...childSnapshot.val() 
        });
      });
      dispatch(setArticles(articles));
    });
  }
}