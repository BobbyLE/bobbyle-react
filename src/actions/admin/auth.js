import { firebase, googleAuthProvider } from '../../firebase/firebase';

export const login = ({uid, displayName, email, photoURL}) => ({
  type: 'LOGIN',
  uid,
  displayName,
  email,
  photoURL
});

export const startLogin = () => {
  return () => {
    return firebase.auth().signInWithPopup(googleAuthProvider);
  };
};

export const startLoginWithEmail = (email, password) => {
  return () => {
    return firebase.auth().signInAndRetrieveDataWithEmailAndPassword(email, password);
  }
}
export const logout = () => ({
  type: 'LOGOUT'
});

export const startLogout = () => {
  return () => {
    return firebase.auth().signOut();
  };
};
