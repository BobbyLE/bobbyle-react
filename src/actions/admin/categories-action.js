import uuid from 'uuid';
import database from '../../firebase/firebase';

//ADD_CATEGORIES
export const addCategory = (category) => ({
  type: 'ADD_CATEGORY',
  category
});

export const startAddCategory = (categoryData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      name = ''
    } = categoryData;
    const category = { name };

    return database.ref(`users/${uid}/categories`).push(category).then((ref) => {
      dispatch(addCategory({
        id: ref.key,
        ...category
      }));
    });
  };
};

// REMOVE_CATEGORY
export const removeCategory = ({ id } = {}) => ({
  type: 'REMOVE_CATEGORY',
  id
});
export const startRemoveCategory = ({id} = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).remove().then(()=> {
      dispatch(removeCategory({ id }));
    });
  } 
};


// EDIT_CATEGORY
export const editCategory =  (id, updates) => ({
  type: 'EDIT_CATEGORY',
  id,
  updates
})

export const startEditCategory = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories/${id}`).update(updates).then(() => {
      return dispatch(editCategory(id, updates));
    });
  };
};

//SET_CATEGORIES
export const setCategories = (categories) => ({
  type: 'SET_CATEGORIES',
  categories
});

export const startSetCategories = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/categories`).once('value').then( (snapshot) => {
      const categories = [];
      snapshot.forEach( (childSnapshot) => {
        categories.push({
          id: childSnapshot.key,
          ...childSnapshot.val() 
        });
      });
      dispatch(setCategories(categories));
    });
  }
}