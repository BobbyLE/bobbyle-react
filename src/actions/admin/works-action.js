import uuid from 'uuid';
import database, { storage } from '../../firebase/firebase';

//ADD_CATEGORIES
export const addWork = (work) => ({
  type: 'ADD_WORK',
  work
});

export const startAddWork = (workData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      tags = '',
      img = '',
      imgURL = ''
    } = workData;
    const work = { title, tags, img, imgURL };

    return database.ref(`users/${uid}/works`).push(work).then((ref) => {
      dispatch(addWork({
        id: ref.key,
        ...work
      }));
    });
  };
};

// REMOVE_CATEGORY
export const removeWork = ({ id } = {}) => ({
  type: 'REMOVE_WORK',
  id
});
export const startRemoveWork = ({id, img} = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    storage.ref(`works/${img}`).delete().catch((error) => {
      console.log(error);
    });
    return database.ref(`users/${uid}/works/${id}`).remove().then(()=> {
      dispatch(removeWork({ id }));
    });
  } 
};


// EDIT_WORK
export const editWork =  (id, updates) => ({
  type: 'EDIT_WORK',
  id,
  updates
})

export const startEditWork = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/works/${id}`).update(updates).then(() => {
      return dispatch(editWork(id, updates));
    });
  };
};

//SET_WORKS
export const setWorks = (works) => ({
  type: 'SET_WORKS',
  works
});

export const startSetWorks = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid ? getState().auth.uid : process.env.USER_ID;
    return database.ref(`users/${uid}/works`).once('value').then( (snapshot) => {
      const works = [];
      snapshot.forEach( (childSnapshot) => {
        works.push({
          id: childSnapshot.key,
          ...childSnapshot.val() 
        });
      });
      dispatch(setWorks(works));
    });
  }
}