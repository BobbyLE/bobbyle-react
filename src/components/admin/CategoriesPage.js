import React from 'react';

import PageHeader from './PageHeader';
import CategoryList from './CategoryList';
import AddCategory from './AddCategory';

export class CategoriesPage extends React.Component  {
  render() {
    return (
      <div>
        <PageHeader title="Categories"/>
        <CategoryList />
        <div className="content-container">
          <AddCategory />
        </div>
      </div>
    );
  }
}

export default CategoriesPage;