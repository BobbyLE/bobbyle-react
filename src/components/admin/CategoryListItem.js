import React from 'react';
import { Link } from 'react-router-dom';


const CategoryListItem = ({id, name}, props) =>  (
  <Link className="list-item" to={`/admin/categories/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{name}</h3>
    </div>
  </Link>
)
export default CategoryListItem;