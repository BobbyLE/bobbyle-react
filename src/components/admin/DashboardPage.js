import React from 'react';
import { connect } from 'react-redux';

export const DashboardPage = (props) => {
  return (
    <div>
      Welcome {props.currentUser.email}
    </div>
  );
};

const mapStateToProps = (state) => ({
  currentUser: state.auth
});

export default connect(mapStateToProps)(DashboardPage);
