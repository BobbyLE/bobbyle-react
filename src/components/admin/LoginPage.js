import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { startLogin, startLoginWithEmail } from '../../actions/admin/auth';
import LoginForm from './LoginForm';

export class LoginPage extends React.Component {
  static propTypes = {
    startLoginWithEmail: PropTypes.func
  }
  static defaultProps = {
    startLoginWithEmail: (() => {})
  }
  state = {
    email: '',
    password: '',
    error: ''
  }
  handleForm = (event) => {
    event.preventDefault();
    if(!this.state.email) {
      this.setState( () => ({ error: 'Please provide an Email.' }));
    } else if(this.state.email !== this.state.email.trim()) {
      this.setState( () => ({ error: 'Please provide a valid Email.' }));
    } else if(!this.state.password) {
      this.setState( () => ({ error: 'Please provide a password.' }));
    } 
    else if(this.state.password !== this.state.password.trim()) {
      this.setState( () => ({ error: 'Please provide a valid password' }));
    }
    else {
      this.setState(()=> ({error: ''}));
      this.props.startLoginWithEmail(this.state.email, this.state.password)
        .then(() => {
          this.setState(()=> ({
            error: '',
            email: '',
            password: ''
          }));
        })
        .catch(function(error) {
          var errorCode = error.code;
          var errorMessage = error.message;
          this.setState(()=> ({error:error.message}));
        }.bind(this));
    }
  }
  onEmailChange = (event) => {
    const email = event.target.value
    this.setState(()=>({ email }));
  }
  onPasswordChange = (event) => {
    const password = event.target.value
    this.setState(()=>({ password }));
  }
  
  render() {
    return (
      <div className="box-layout">
      <div className="box-layout__box">
        <h1 className="box-layout__title">Bobby Le</h1>
        {this.state.error && <p className="add-option-error">{this.state.error}</p>}
        <form className="form" onSubmit={this.handleForm}>
          <input 
            type="text" 
            className="text-input" 
            placeholder="email" 
            onChange={this.onEmailChange}
            value={this.state.email}          
          />
          <input 
            type="password" 
            className="text-input" 
            placeholder="Password"
            onChange={this.onPasswordChange}
            value={this.state.password}
          />
          <button className="button">Login</button>
        </form>
      </div>
    </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startLoginWithEmail: (email, password) => dispatch(startLoginWithEmail(email, password))
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
