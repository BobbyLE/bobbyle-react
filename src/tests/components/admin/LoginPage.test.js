import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../../components/admin/LoginPage';

test('should correctly render LoginPage', () => {
  const wrapper = shallow(<LoginPage />);
  expect(wrapper).toMatchSnapshot()
});