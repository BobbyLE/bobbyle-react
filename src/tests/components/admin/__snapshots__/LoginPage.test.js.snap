// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render LoginPage 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      Bobby Le
    </h1>
    <form
      className="form"
      onSubmit={[Function]}
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="email"
        type="text"
        value=""
      />
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        className="button"
      >
        Login
      </button>
    </form>
  </div>
</div>
`;
